### PR TITLE
Add ability to filter GraphViewer using IndexedSubGraph. 

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Add ability to filter GraphViewer using IndexedSubGraph. 

**Motivation and Context**
This is to support compiling execution providers in a minimal build where we need to call IExecutionProvider::Compile but don't have the ability to create a new Graph instance from the IndexedSubGraph returned by IExecutionProvider::GetCapability. Instead we can filter the main Graph instance using IndexedSubGraph to pass the same information through when calling Compile.

### CODE REVIEW

File: include/onnxruntime/core/graph/graph_nodes.h [hunk 1]
```diff
@@ -48,19 +54,25 @@ class ValidNodes {
...
   }
 
-  MutableNodeIterator begin() noexcept {
-    return {nodes_.begin(), nodes_.end()};
+  // we only allow mutable access if the container is non-const.
+  // we need to templatize the functions for enable_if to work at this level, but mandate T2 being TNodesContainer
+  template <typename T2 = TNodesContainer>
+  typename std::enable_if<!std::is_const<T2>::value, MutableNodeIterator>::type begin() noexcept {
+    static_assert(std::is_same<T2, TNodesContainer>::value, "Explicit specialization is not allowed");
+    return MutableNodeIterator(nodes_.begin(), nodes_.end(), filter_node_fn_);
```
Comment (gwang-msft) [inline include/onnxruntime/core/graph/graph_nodes.h:69]: Do you actually need this static_assert? Isn't T2=TNodesContainer already?

File: include/onnxruntime/core/graph/graph_nodes.h [hunk 2]
```diff
@@ -132,18 +147,36 @@ class ValidNodes {
    private:
     TIterator current_;
     TIterator end_;
+    bool filter_;                        // store whether filter_func_ is not nullptr and contains a callable
```
Comment (gwang-msft) [inline include/onnxruntime/core/graph/graph_nodes.h:150]: >filter_ [](start = 9, length = 7)

has_filter_?  filter_, filter_func_ and filter_node_fn_ make it very confusing

File: onnxruntime/core/graph/graph_viewer.cc [hunk 3]
```diff
@@ -61,23 +77,75 @@ GraphViewer::GraphViewer(const Graph& graph) {
...
       },
       PriorityNodeCompare());
+#endif
+
+  if (filter_info_) {
+    // create set of node indexes as we need quick lookups and don't care about the order
+    filtered_node_indices_ = std::unordered_set<NodeIndex>(filter_info->nodes.cbegin(),
+                                                           filter_info->nodes.cend());
+
+    // validate. if something is off here it's a bug in our code
```
Comment (gwang-msft) [inline onnxruntime/core/graph/graph_viewer.cc:93]: Verify first then copy?

File: include/onnxruntime/core/graph/graph_nodes.h (hunk 1 above)
Comment (skottmckay) [reply include/onnxruntime/core/graph/graph_nodes.h:69]: It's defaulted to TNodesContainer. the static_assert enforces at compile time that nobody tries to override that.

File: include/onnxruntime/core/graph/graph_nodes.h (hunk 2 above)
Comment (skottmckay) [reply include/onnxruntime/core/graph/graph_nodes.h:150]: Renamed to apply_filter_ and made the init `apply_filter_{filter_fn != nullptr}` to try and make it clearer

File: onnxruntime/test/ir/graph_viewer_test.cc [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  subgraph.SetMetaDef(std::move(metadef));
+
+  GraphViewer viewer{graph, subgraph};
+
+  EXPECT_EQ(viewer.NumberOfNodes(), 2);
+  // we added nodes in the topo order so that should still be the case
+  EXPECT_THAT(subgraph.nodes, testing::ContainerEq(viewer.GetNodesInTopologicalOrder()));
+
+  nodes = graph.Nodes().cbegin();
+  auto end_nodes = graph.Nodes().cend();
```
Comment (gwang-msft) [inline onnxruntime/test/ir/graph_viewer_test.cc:76]: node and end_node?

**Review by gwang-msft — APPROVED** [review]
:shipit:

File: onnxruntime/test/ir/graph_viewer_test.cc (hunk 4 above)
Comment (skottmckay) [reply onnxruntime/test/ir/graph_viewer_test.cc:76]: Renamed `nodes` to `cur_node`. Left as `end_nodes` as it's the iterator pointing to the end of the nodes marker and not the last node.

**Review by gwang-msft — APPROVED** [review]
:shipit:

**Pull request merged**